sherpa/utils/__init__.py

```diff
@@ -4188,6 +4188,59 @@ def send_to_pager(txt, filename=None, clobber=False):
         print(txt, file=fh)
 
 
+class ObjectStore:
```

Comment:
So, I have a tree of data classes - e.g. Data1D, Data1DInt, DataPHA - and potentially different values to store for each class. I then come along and say given an instance of one of these classes, what is the appropriate value from the store? The Data1DInt and DataPHA classes are derived from Data1D but they are not derived from each other, and it may be that DataPHA stores a value but Data1DInt doesn't.

We use the MRO depth to determine what value is a good match. This could well be over-engineering...

Reply:
I had to make a similar design decision for a different package a while ago (https://github.com/Chandra-MARX/marxs/blob/main/marxs/visualization/utils.py#L27). There is a similar hierarchy of classes and also the need to set per-object preferences. So, what I settled on there was to has a dictionary for each object and a special look-up so that any key that is not set in that object is looked up in the class dictionary, and if it's not found there, it's checked in the base class etc. That's similar to what you propose here and I found it to work. That way, one can, e.g. change the line width for just one histogram plot, without changing it for all current and future plots. But one can also (by modifying the class dictionary) change all histogram plots (except for those that override the setting in their own instance dictionary. How to make this transparent on the UI layer is a different question...